src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1532,6 +1581,47 @@ private EmitResult ToEmitResultAndFree(DiagnosticBag diagnostics, bool success)
 
         internal bool SerializeToPeStream(
             CommonPEModuleBuilder moduleBeingBuilt,
+            EmitStreamProvider emitStreamProvider,
+            string pdbFileName,
+            Func<object> testSymWriterFactory,
+            DiagnosticBag diagnostics,
+            bool metadataOnly,
+            CancellationToken cancellationToken)
+        {
+            Stream peStream = emitStreamProvider.GetPeStream(diagnostics);
+            if (peStream == null)
+            {
+                Debug.Assert(diagnostics.HasAnyErrors());
+                return false;
+            }
+            Debug.Assert(peStream.CanWrite);
+
+            Stream pdbStream = null;
+            if (emitStreamProvider.HasPdbStream)
+            {
+                pdbStream = emitStreamProvider.GetPdbStream(diagnostics);
+                if (pdbStream == null)
+                {
+                    Debug.Assert(diagnostics.HasAnyErrors());
+                    return false;
+                }
+
+                Debug.Assert(pdbStream.CanWrite);
+            }
+
+            return SerializeToPeStream(
```

Comment:
I worry that we might encounter some errors even after this point, but before anything is written to the streams. There is a reason we pass DiagnosticBag to this method. Consider requesting particular stream right before we start writing to it, perhaps somewhere in PEWriter.